rdt/transformers/id.py

```diff
@@ -172,27 +183,43 @@ def _generate_fallback_samples(self, num_samples, template_samples):
             # samples as a base and appends a counter to make each value unique.
             counter = 0
             samples = []
-            while num_samples > 0:
+            while num_samples > len(samples):
                 samples.extend([f'{i}({counter})' for i in template_samples[:num_samples]])
-                num_samples -= len(template_samples)
                 counter += 1
 
-            return samples
+            return samples[:num_samples]
 
     def _generate_unique_regex_values(self):
         regex_values = []
         try:
             while len(regex_values) < self._data_cardinality:
                 regex_values.append(next(self.generator))
         except (RuntimeError, StopIteration):
-            fallback_samples = self._generate_fallback_samples(
+            fallback_samples = self._sample_fallback(
                 num_samples=self._data_cardinality - len(regex_values),
                 template_samples=regex_values,
             )
             regex_values.extend(fallback_samples)
 
         return regex_values
 
+    def _get_cardinality_frequency(self, data):
+        """Get number of repetitions and their frequencies."""
+        repetitions = data.value_counts().to_numpy()
+        num_nans = sum(pd.isna(data))
+        if num_nans > 0:
+            repetitions = np.concatenate([repetitions, [num_nans]])
+
+        repetitions_counter = Counter(repetitions)
+        total_repetitions = sum(repetitions_counter.values())
+        normalized_frequency = {
+            repetition: freq / total_repetitions for repetition, freq in repetitions_counter.items()
+        }
+        sorted_repetitions = sorted(normalized_frequency.keys())
+        sorted_frequencies = [normalized_frequency[repetition] for repetition in sorted_repetitions]
+
+        return sorted_repetitions, sorted_frequencies
```

Comment:
I think that this could be done with pandas in a more easy way. I'm not 100% sure about the 'nan' and 'None's but I'm thinking of something like this:
```
    value_counts = data.value_counts(dropna=False)

    repetition_counts = value_counts.value_counts().sort_index()
    total = repetition_counts.sum()

    # Since this is pandas we can apply this in one go
    frequencies = (repetition_counts / total).to_numpy()
    repetitions = repetition_counts.index.to_numpy()
```

Reply:
Thank you for the suggestion. The nans can be handled with the line `data = fill_nan_with_none(data)`.